File: python/year_2023/day_06.py
from pathlib import Path
from math import isqrt, prod


def part_1(input: str) -> int:
    times, _, distances = input.rstrip().partition("\n")
    times = map(int, times.strip("Time:").split())
    distances = map(int, distances.strip("Distance:").split())
    return prod(map(count_ways_to_beat, times, distances))


def part_2(input: str) -> int:
    time, _, distance = input.rstrip().partition("\n")
    time = int(time.strip("Time:").replace(" ", ""))
    distance = int(distance.strip("Distance:").replace(" ", ""))
    return count_ways_to_beat(time, distance)


def count_ways_to_beat(time: int, distance: int) -> int:
    sqrt_discriminant = isqrt(time * time - 4 * distance - 1) + 1
    i = (-time + sqrt_discriminant) // -2 + 1
    j = ceil_div(-time - sqrt_discriminant, -2)
    return j - i


def ceil_div(a: int, b: int) -> int:
    return -(a // -b)




def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_06.txt").read_text()

def test_part_1_example_1():
    assert part_1(EXAMPLE_1) == 288


def test_part_1_input():
    assert part_1(_read_input()) == 625968


def test_part_2_example_1():
    assert part_2(EXAMPLE_1) == 71503


def test_part_2_input():
    assert part_2(_read_input()) == 43663323


EXAMPLE_1 = """Time:      7  15   30
Distance:  9  40  200"""

File: rust/year_2023/src/day_06.rs
pub fn part_1(input: &str) -> isize {
    let (times, distances) = input.trim_end().split_once("\n").unwrap();
    let times = times
        .strip_prefix("Time:")
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse::<isize>().unwrap());
    let distances = distances
        .strip_prefix("Distance:")
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse::<isize>().unwrap());
    times
        .zip(distances)
        .map(|(t, d)| count_ways_to_beat(t, d))
        .product()
}

pub fn part_2(input: &str) -> isize {
    let (time, distance) = input.trim_end().split_once("\n").unwrap();
    let time = time
        .strip_prefix("Time:")
        .unwrap()
        .replace(" ", "")
        .parse::<isize>()
        .unwrap();
    let distance = distance
        .strip_prefix("Distance:")
        .unwrap()
        .replace(" ", "")
        .parse::<isize>()
        .unwrap();
    count_ways_to_beat(time, distance)
}

fn count_ways_to_beat(time: isize, distance: isize) -> isize {
    let sqrt_discriminant = (time * time - 4 * distance - 1).isqrt() + 1;
    let i = (-time + sqrt_discriminant) / -2 + 1;
    let j = (-time - sqrt_discriminant).div_ceil(-2);
    return j - i;
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../../../testdata/year_2023/day_06.txt");

    const EXAMPLE_1: &str = "Time:      7  15   30
Distance:  9  40  200";

    #[test_case(EXAMPLE_1, 288)]
    #[test_case(INPUT, 625968)]
    fn part_1(input: &str, expected: isize) {
        assert_eq!(super::part_1(input), expected);
    }

    #[test_case(EXAMPLE_1, 71503)]
    #[test_case(INPUT, 43663323)]
    fn part_2(input: &str, expected: isize) {
        assert_eq!(super::part_2(input), expected);
    }
}
